src/main/scala/com/microsoft/azure/cosmosdb/spark/CosmosDBConnection.scala

```diff
@@ -22,281 +22,77 @@
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryAttemptsOnThrottledRequests(0);
-      documentClient.getConnectionPolicy().getRetryOptions().setMaxRetryWaitTimeInSeconds(0);
-    }
+  def getCollectionLink: String = {
+    collectionLink
   }
 
-  def getCollection: DocumentCollection = {
-    if (collection == null) {
-      collection = documentClient.readCollection(collectionLink, null).getResource
-    }
-    collection
-  }
-
-  def getDatabase: Database = {
-    if (database == null) {
-      database = documentClient.readDatabase(databaseLink, null).getResource
-    }
-    database
+  def reinitializeClient(): AnyVal = {
+    CosmosDBConnectionCache.reinitializeClient(clientConfig)
   }
 
   def getAllPartitions: Array[PartitionKeyRange] = {
-    var ranges = documentClient.readPartitionKeyRanges(collectionLink, null.asInstanceOf[FeedOptions])
+    val documentClient = CosmosDBConnectionCache.getOrCreateClient(clientConfig)
+    val ranges = documentClient.readPartitionKeyRanges(collectionLink, null.asInstanceOf[FeedOptions])
     ranges.getQueryIterator.toArray
   }
 
-  def getAllPartitions(query: String): Array[PartitionKeyRange] = {
-    getAllPartitions
+  def getDocumentBulkImporter: DocumentBulkExecutor = {
+    CosmosDBConnectionCache.getOrCreateBulkExecutor(clientConfig)
   }
 
-  def getCollectionThroughput: Int = {
-    collectionThroughput match {
-      case Some(value) => value
-      case None => {
-        var offers = documentClient.queryOffers(s"SELECT * FROM c where c.offerResourceId = '${getCollection.getResourceId}'", null).getQueryIterable.toList
-        if (offers.isEmpty) {
-          offers = documentClient.queryOffers(s"SELECT * FROM c where c.offerResourceId = '${getDatabase.getResourceId}'", null).getQueryIterable.toList
-          // database throughput
-          if (offers.isEmpty) {
-            throw new IllegalStateException("Cannot find the collection corresponding offer.")
-          }
-        }
-
-        val offer = offers.get(0)
-        val collectionThroughput = if (offer.getString("offerVersion") == "V1")
-          CosmosDBConfig.SinglePartitionCollectionOfferThroughput
-        else
-          offer.getContent.getInt("offerThroughput")
-
-        this.collectionThroughput = Some(collectionThroughput)
-
-        logInfo(s"getCollectionThroughput: ${collectionThroughput}")        
-        collectionThroughput
-      }
-    }  
+  def getPartitionKeyDefinition: PartitionKeyDefinition = {
+    CosmosDBConnectionCache.getPartitionKeyDefinition(clientConfig)
   }
 
-  def reinitializeClient (shouldReinitializeThroughput: Boolean) = {
-    documentClient = CosmosDBConnection.reinitializeClient(getCollection, connectionMode, getClientConfiguration(config))
-
-    if (shouldReinitializeThroughput) {
-      logInfo("reinitializeClient: Resetting cached collection throughput")
-      collectionThroughput = None
-    }
-  }
+  def queryDocuments(
```

Comment:
I saw this in a few places but I prefer to have the first parameter on the same line as method name, and for methods with a lot of parameters, they're on new lines and aligned.